File: src/reflection_memory/historical_memory.py
import faiss
import numpy as np
from typing import List, Tuple, Dict, Any
from sklearn.feature_extraction.text import TfidfVectorizer
from .models import IncidentReport

class HistoricalMemoryAgent:
    """
    HistoricalMemoryAgent provides RAG capability over past incident reports using FAISS.
    It can be queried mid-reasoning at any stage of negotiation to fetch past decisions,
    failures, and mitigation lessons learned.
    """

    def __init__(self, agent_id: str = "HistoricalMemory_01"):
        self.agent_id = agent_id
        self.incidents: List[IncidentReport] = []
        self.vectorizer = TfidfVectorizer(stop_words='english', max_features=512)
        self.index: faiss.IndexFlatIP = None  # Inner product index for cosine similarity
        self.is_fitted = False
        self._seed_default_incidents()

    def _seed_default_incidents(self):
        """Seed the memory bank with realistic system failure & negotiation incident reports."""
        default_incidents = [
            IncidentReport(
                id="INC-2024-089",
                title="Monolith Cloud Migration Memory Leak",
                domain="Infrastructure",
                description="Migrated core service without worker thread pool limits under 4x traffic peak. Memory footprint spiked, causing cascading pod crashes.",
                outcome="FAILURE: System suffered 42-minute global outage.",
                risk_score=0.88,
                lessons_learned=[
                    "Enforce strict per-container memory limits.",
                    "Mandate auto-scaling policies with automatic failover.",
                    "Never deploy capacity changes without load-test proof."
                ],
                tags=["infrastructure", "traffic_spike", "memory_leak", "failover", "outage"]
            ),
            IncidentReport(
                id="INC-2025-014",
                title="Uncapped Marketing Budget Allocation",
                domain="Finance",
                description="Approved $120k ad campaign budget without dynamic spend safety limits or ROI benchmarks during Q2 launch.",
                outcome="PARTIAL LOSS: Overspent by $45k with sub-1.0 ROI.",
                risk_score=0.75,
                lessons_learned=[
                    "Cap total unhedged campaign budget at $50,000.",
                    "Require SafetyAgent SOP approval for all tier-1 budget changes."
                ],
                tags=["budget", "finance", "overspend", "sop_violation"]
            ),
            IncidentReport(
                id="INC-2025-042",
                title="Database Deadlock under High Concurrent Write Rate",
                domain="Database",
                description="Disabled read-replicas during peak flash sale to save infrastructure costs, causing database lock contention.",
                outcome="CRITICAL: Transaction failure rate hit 68%.",
                risk_score=0.92,
                lessons_learned=[
                    "Read-replicas must remain active during any high-concurrency event.",
                    "Blue-green deployment strategy required for DB schema updates."
                ],
                tags=["database", "deadlock", "concurrency", "traffic_spike", "replicas"]
            ),
            IncidentReport(
                id="INC-2025-103",
                title="Successful Auto-Scale Rollout with Fallback Strategy",
                domain="Infrastructure",
                description="Deployed gradual blue-green rollout with 20% canary traffic, automated fallback triggers, and active load monitoring.",
                outcome="SUCCESS: Handled 5x peak traffic with 99.99% availability.",
                risk_score=0.15,
                lessons_learned=[
                    "Canary deployments reduce blast radius effectively.",
                    "Pre-warmed instances mitigate latency spikes during scaling."
                ],
                tags=["canary", "auto_scaling", "success", "blue_green", "fallback"]
            )
        ]
        for inc in default_incidents:
            self.add_incident(inc)
        self.build_index()

    def add_incident(self, incident: IncidentReport):
        self.incidents.append(incident)
        self.is_fitted = False

    def build_index(self):
        """Build FAISS vector index over all stored incident descriptions and tags."""
        if not self.incidents:
            return

        corpus = [
            f"{inc.title} {inc.domain} {inc.description} {' '.join(inc.tags)} {' '.join(inc.lessons_learned)}"
            for inc in self.incidents
        ]

        tfidf_matrix = self.vectorizer.fit_transform(corpus).toarray().astype(np.float32)
        # Normalize vectors for Cosine Similarity using Inner Product index
        faiss.normalize_L2(tfidf_matrix)

        dimension = tfidf_matrix.shape[1]
        self.index = faiss.IndexFlatIP(dimension)
        self.index.add(tfidf_matrix)
        self.is_fitted = True

    def query_mid_reasoning(self, query: str, top_k: int = 2) -> List[Tuple[IncidentReport, float]]:
        """
        Queryable mid-reasoning at any point during negotiation loop.
        Returns list of (IncidentReport, similarity_score) tuples.
        """
        if not self.is_fitted or self.index is None:
            self.build_index()

        query_vec = self.vectorizer.transform([query]).toarray().astype(np.float32)
        faiss.normalize_L2(query_vec)

        scores, indices = self.index.search(query_vec, min(top_k, len(self.incidents)))

        results = []
        for idx, score in zip(indices[0], scores[0]):
            if idx >= 0 and idx < len(self.incidents):
                results.append((self.incidents[idx], float(score)))

        return results

File: src/reflection_memory/historical_memory.ts
import { IncidentReport } from "./models.js";

export class HistoricalMemoryAgent {
  public agentId: string;
  public incidents: IncidentReport[] = [];
  private vocab: Map<string, number> = new Map();
  private idf: number[] = [];
  private indexVectors: number[][] = [];
  private isFitted: boolean = false;

  constructor(agentId: string = "HistoricalMemory_01") {
    this.agentId = agentId;
    this._seedDefaultIncidents();
  }

  private _seedDefaultIncidents(): void {
    const defaultIncidents: IncidentReport[] = [
      {
        id: "INC-2024-089",
        title: "Machine #4 Bearing Failure & Thermal Runaway",
        domain: "Plant Maintenance",
        description: "Operated Machine #4 CNC spindle without lubrication monitoring under 4x production throughput peak. Thermal expansion caused bearing lockup and conveyor line jam.",
        outcome: "FAILURE: Production line suffered 42-minute factory floor shutdown.",
        risk_score: 0.88,
        lessons_learned: [
          "Enforce strict per-machine thermal and vibration limits.",
          "Mandate predictive maintenance policies with automatic safety interlocks.",
          "Never deploy production speed increases without load-test proof."
        ],
        tags: ["manufacturing", "machine_4", "bearing_failure", "interlock", "shutdown"]
      },
      {
        id: "INC-2025-014",
        title: "Uncapped CNC Tooling Budget Allocation",
        domain: "Plant Operations",
        description: "Approved $120k tooling upgrade budget without dynamic spend safety limits or ROI benchmarks during Q2 production run.",
        outcome: "PARTIAL LOSS: Overspent by $45k with sub-1.0 ROI.",
        risk_score: 0.75,
        lessons_learned: [
          "Cap total unhedged tooling budget at $50,000.",
          "Require SafetyAgent SOP approval for all tier-1 capital equipment changes."
        ],
        tags: ["budget", "plant_operations", "overspend", "sop_violation"]
      },
      {
        id: "INC-2025-042",
        title: "PLC Lockup under High Concurrent Conveyor Feed Rate",
        domain: "Automation / PLC",
        description: "Disabled redundant sensor checks during peak production run to save power costs, causing PLC input lock contention.",
        outcome: "CRITICAL: Robot arm assembly failure rate hit 68%.",
        risk_score: 0.92,
        lessons_learned: [
          "Redundant PLC sensor feeds must remain active during any high-throughput event.",
          "Staged rollout strategy required for PLC firmware and sensor updates."
        ],
        tags: ["plc", "sensor_drift", "conveyor", "high_throughput", "redundancy"]
      },
      {
        id: "INC-2025-103",
        title: "Successful Machine #4 Predictive Maintenance & Interlock Rollout",
        domain: "Predictive Maintenance",
        description: "Deployed gradual staged rollout on Machine #4 with 20% conveyor load, automated interlock triggers, and active sensor monitoring.",
        outcome: "SUCCESS: Handled 5x production throughput with 99.99% operational uptime.",
        risk_score: 0.15,
        lessons_learned: [
          "Staged rollouts reduce blast radius on assembly lines effectively.",
          "Pre-calibrated sensors mitigate calibration drift during scaling."
        ],
        tags: ["staged_rollout", "predictive_maintenance", "success", "machine_4", "interlock"]
      }
    ];

    for (const inc of defaultIncidents) {
      this.addIncident(inc);
    }
    this.buildIndex();
  }

  public addIncident(incident: IncidentReport): void {
    this.incidents.push(incident);
    this.isFitted = false;
  }

  private _tokenize(text: string): string[] {
    return text
      .toLowerCase()
      .replace(/[^a-z0-9\s]/g, " ")
      .split(/\s+/)
      .filter((w) => w.length > 2);
  }

  public buildIndex(): void {
    if (this.incidents.length === 0) return;

    const documents = this.incidents.map((inc) =>
      `${inc.title} ${inc.domain} ${inc.description} ${inc.tags.join(" ")} ${inc.lessons_learned.join(" ")}`
    );

    const tokenizedDocs = documents.map((doc) => this._tokenize(doc));

    // Build vocabulary
    this.vocab.clear();
    let termId = 0;
    for (const tokens of tokenizedDocs) {
      for (const t of tokens) {
        if (!this.vocab.has(t)) {
          this.vocab.set(t, termId++);
        }
      }
    }

    const vocabSize = this.vocab.size;
    const numDocs = documents.length;

    // Calculate IDF
    const docFreq = new Array(vocabSize).fill(0);
    for (const tokens of tokenizedDocs) {
      const uniqueTokens = new Set(tokens);
      for (const t of uniqueTokens) {
        const id = this.vocab.get(t);
        if (id !== undefined) {
          docFreq[id]++;
        }
      }
    }

    this.idf = docFreq.map((df) => Math.log((numDocs + 1) / (df + 1)) + 1.0);

    // Compute TF-IDF vectors for index
    this.indexVectors = tokenizedDocs.map((tokens) => {
      const vec = new Array(vocabSize).fill(0);
      const tf = new Map<string, number>();
      for (const t of tokens) {
        tf.set(t, (tf.get(t) || 0) + 1);
      }
      for (const [t, count] of tf.entries()) {
        const id = this.vocab.get(t);
        if (id !== undefined) {
          vec[id] = (count / tokens.length) * this.idf[id];
        }
      }
      // L2 Normalize
      const norm = Math.sqrt(vec.reduce((sum, v) => sum + v * v, 0));
      return norm > 0 ? vec.map((v) => v / norm) : vec;
    });

    this.isFitted = true;
  }

  public queryMidReasoning(query: string, topK: number = 2): [IncidentReport, number][] {
    if (!this.isFitted || this.indexVectors.length === 0) {
      this.buildIndex();
    }

    const queryTokens = this._tokenize(query);
    const vocabSize = this.vocab.size;
    const queryVec = new Array(vocabSize).fill(0);

    const tf = new Map<string, number>();
    for (const t of queryTokens) {
      tf.set(t, (tf.get(t) || 0) + 1);
    }
    for (const [t, count] of tf.entries()) {
      const id = this.vocab.get(t);
      if (id !== undefined) {
        queryVec[id] = (count / queryTokens.length) * this.idf[id];
      }
    }

    // L2 Normalize query vector
    const norm = Math.sqrt(queryVec.reduce((sum, v) => sum + v * v, 0));
    const normalizedQuery = norm > 0 ? queryVec.map((v) => v / norm) : queryVec;

    // Calculate Cosine Similarity scores
    const scored: [IncidentReport, number][] = this.incidents.map((inc, i) => {
      const docVec = this.indexVectors[i];
      let dotProduct = 0;
      for (let j = 0; j < vocabSize; j++) {
        dotProduct += normalizedQuery[j] * docVec[j];
      }
      return [inc, dotProduct];
    });

    // Sort descending by similarity score
    scored.sort((a, b) => b[1] - a[1]);

    return scored.slice(0, Math.min(topK, scored.length));
  }
}
